src/jdk.compiler/share/classes/com/sun/tools/javac/comp/InferenceContext.java

```diff
@@ -243,7 +262,72 @@ List<Type> instTypes() {
      * fully instantiated, it will still be available in the resulting type.
      */
     Type asInstType(Type t) {
-        return types.subst(t, inferencevars, instTypes());
+        Type result = types.subst(t, inferencevars, instTypes());
+        InferenceContext next = supplementaryIC;
+        // stop as soon as there are no more supplementary inference context or
+        // no more type variables to be instantiated
+        if (next != null) {
+            HasTypeVars hasTypeVars = new HasTypeVars();
+            while (next != null && hasTypeVars.visit(result)) {
```

Comment:
I think that we should be fine just going for the next supplementaryIC instead of traversing the whole list here. Same for the process to add free listeners, but in the first version I preferred to err on the correctness / completeness side. If we decide to visit more than one chained supplementary IC we can still do some improvements to try to reduce the number of supplementary ICs visited to a minimum

Reply:
Maybe we can replace the visitor with asking `supplimentaryIC.free(t)` ? e.g. this could maybe use recusion?

```
Type result = t;
if (current.free(t)) {
    result = types.subst(t, inferencevars, instTypes());
}
if (supplimentaryIC != null) {
    return supplimentaryIC.asInstTypes(t);
 }
 ```

Reply:
> Maybe we can replace the visitor with asking `supplimentaryIC.free(t)` ? e.g. this could maybe use recusion?
> 
> ```
> Type result = t;
> if (current.free(t)) {
>     result = types.subst(t, inferencevars, instTypes());
> }
> if (supplimentaryIC != null) {
>     return supplimentaryIC.asInstTypes(t);
>  }
> ```

right but won't this solution imply to recursively visit the whole chain of supplementaryICs? the current approach should stop as soon as there are no more supplementaryICs to visit. Actually I think that one of our performance tests failed and that was why I had to introduce this visitor if I recall correctly

Reply:
> Actually I think that one of our performance tests failed and that was why I had to introduce this visitor if I recall correctly

I wanted to ask about this -- e.g. how much fixing this undoes the benefit of minimization. It seems we're in tricky territory.

Now, the problem with the visitor is that it picks up any type-variables -- it should at least filter out class type-variables and focus on method type-variables.

Reply:
> > Actually I think that one of our performance tests failed and that was why I had to introduce this visitor if I recall correctly
> 
> I wanted to ask about this -- e.g. how much fixing this undoes the benefit of minimization. It seems we're in tricky territory.
> 
> Now, the problem with the visitor is that it picks up any type-variables -- it should at least filter out class type-variables and focus on method type-variables.

the difference in performance is negligible